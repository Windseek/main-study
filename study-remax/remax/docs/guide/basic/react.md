---
title: React
order: 21
---

# React

在使用 Remax 前你需要对 [React](https://reactjs.org/) 有基本的了解。

如果你有 React 的开发经验，那么也要注意 Remax 不同于 React DOM。Remax 使用小程序提供的组件来构建应用，而且你的代码是运行在小程序环境中。

你在 Remax 中使用 React 没有任何特殊限制，同时也支持 ES6 语法， ES Module，async/await 等等。
