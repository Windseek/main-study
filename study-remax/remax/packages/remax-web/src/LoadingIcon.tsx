import * as React from 'react';

interface LoadingIconProps {
  animate?: boolean;
}

const values = '2; 2; 2; 2';
const animateValues = '0; 4; 0; 0';

export default function LoadingIcon(props: LoadingIconProps) {
  const animate = props.animate;

  return (
    <svg className="remax-pull-to-refresh-loading" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32">
      <circle transform="translate(8 0)" cx="0" cy="16" r="0">
        <animate
          attributeName="r"
          values={!animate ? values : animateValues}
          dur="1.2s"
          repeatCount="indefinite"
          begin="0"
          keyTimes="0;0.2;0.7;1"
          keySplines="0.2 0.2 0.4 0.8;0.2 0.6 0.4 0.8;0.2 0.6 0.4 0.8"
          calcMode="spline"
        />
      </circle>
      <circle transform="translate(16 0)" cx="0" cy="16" r="0">
        <animate
          attributeName="r"
          values={!animate ? values : animateValues}
          dur="1.2s"
          repeatCount="indefinite"
          begin="0.3"
          keyTimes="0;0.2;0.7;1"
          keySplines="0.2 0.2 0.4 0.8;0.2 0.6 0.4 0.8;0.2 0.6 0.4 0.8"
          calcMode="spline"
        />
      </circle>
      <circle transform="translate(24 0)" cx="0" cy="16" r="0">
        <animate
          attributeName="r"
          values={!animate ? values : animateValues}
          dur="1.2s"
          repeatCount="indefinite"
          begin="0.6"
          keyTimes="0;0.2;0.7;1"
          keySplines="0.2 0.2 0.4 0.8;0.2 0.6 0.4 0.8;0.2 0.6 0.4 0.8"
          calcMode="spline"
        />
      </circle>
    </svg>
  );
}
