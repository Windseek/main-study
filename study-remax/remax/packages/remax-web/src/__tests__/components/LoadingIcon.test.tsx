import * as React from 'react';
import TestRenderer from 'react-test-renderer';
import LoadingIcon from '../../LoadingIcon';

describe('LoadingIcon', () => {
  it('render correctly', () => {
    const testRenderer = TestRenderer.create(<LoadingIcon />);

    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it('render with animate', () => {
    const testRenderer = TestRenderer.create(<LoadingIcon animate={true} />);

    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
