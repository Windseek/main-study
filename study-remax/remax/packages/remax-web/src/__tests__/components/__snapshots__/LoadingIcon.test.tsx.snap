// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingIcon render correctly 1`] = `
<svg
  className="remax-pull-to-refresh-loading"
  viewBox="0 0 32 32"
  xmlns="http://www.w3.org/2000/svg"
>
  <circle
    cx="0"
    cy="16"
    r="0"
    transform="translate(8 0)"
  >
    <animate
      attributeName="r"
      begin="0"
      calcMode="spline"
      dur="1.2s"
      keySplines="0.2 0.2 0.4 0.8;0.2 0.6 0.4 0.8;0.2 0.6 0.4 0.8"
      keyTimes="0;0.2;0.7;1"
      repeatCount="indefinite"
      values="2; 2; 2; 2"
    />
  </circle>
  <circle
    cx="0"
    cy="16"
    r="0"
    transform="translate(16 0)"
  >
    <animate
      attributeName="r"
      begin="0.3"
      calcMode="spline"
      dur="1.2s"
      keySplines="0.2 0.2 0.4 0.8;0.2 0.6 0.4 0.8;0.2 0.6 0.4 0.8"
      keyTimes="0;0.2;0.7;1"
      repeatCount="indefinite"
      values="2; 2; 2; 2"
    />
  </circle>
  <circle
    cx="0"
    cy="16"
    r="0"
    transform="translate(24 0)"
  >
    <animate
      attributeName="r"
      begin="0.6"
      calcMode="spline"
      dur="1.2s"
      keySplines="0.2 0.2 0.4 0.8;0.2 0.6 0.4 0.8;0.2 0.6 0.4 0.8"
      keyTimes="0;0.2;0.7;1"
      repeatCount="indefinite"
      values="2; 2; 2; 2"
    />
  </circle>
</svg>
`;

exports[`LoadingIcon render with animate 1`] = `
<svg
  className="remax-pull-to-refresh-loading"
  viewBox="0 0 32 32"
  xmlns="http://www.w3.org/2000/svg"
>
  <circle
    cx="0"
    cy="16"
    r="0"
    transform="translate(8 0)"
  >
    <animate
      attributeName="r"
      begin="0"
      calcMode="spline"
      dur="1.2s"
      keySplines="0.2 0.2 0.4 0.8;0.2 0.6 0.4 0.8;0.2 0.6 0.4 0.8"
      keyTimes="0;0.2;0.7;1"
      repeatCount="indefinite"
      values="0; 4; 0; 0"
    />
  </circle>
  <circle
    cx="0"
    cy="16"
    r="0"
    transform="translate(16 0)"
  >
    <animate
      attributeName="r"
      begin="0.3"
      calcMode="spline"
      dur="1.2s"
      keySplines="0.2 0.2 0.4 0.8;0.2 0.6 0.4 0.8;0.2 0.6 0.4 0.8"
      keyTimes="0;0.2;0.7;1"
      repeatCount="indefinite"
      values="0; 4; 0; 0"
    />
  </circle>
  <circle
    cx="0"
    cy="16"
    r="0"
    transform="translate(24 0)"
  >
    <animate
      attributeName="r"
      begin="0.6"
      calcMode="spline"
      dur="1.2s"
      keySplines="0.2 0.2 0.4 0.8;0.2 0.6 0.4 0.8;0.2 0.6 0.4 0.8"
      keyTimes="0;0.2;0.7;1"
      repeatCount="indefinite"
      values="0; 4; 0; 0"
    />
  </circle>
</svg>
`;
